<a href="https://colab.research.google.com/github/AstronomekDomansky/-.-1/blob/main/%D0%A1%D0%B8%D0%B8%D0%91%D0%942.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# ==========================================================
# ⬇️  ЗАПОЛНИТЕ ЭТИ ТРИ ПОЛЯ СВОИМИ ДАННЫМИ
# ==========================================================
last_name   = "Арутюнян"    # <-- впишите свою фамилию
first_name  = "Дмитрий"      # <-- впишите своё имя
list_number = 2           # <-- ваш номер по списку (целое число)
# ==========================================================

import hashlib

def make_seed(last: str, first: str, number: int) -> int:
    # ФИО + номер -> стабильный seed 0..99999
    key = f"{last.strip().lower()}|{first.strip().lower()}|{int(number)}"
    return int(hashlib.sha256(key.encode("utf-8")).hexdigest(), 16) % 100_000

SEED = make_seed(last_name, first_name, list_number)

print("Студент:", last_name, first_name, f"(№{list_number})")
print("Ваш персональный SEED:", SEED)
print("Запишите SEED в итоговый отчёт — он подтверждает вариант.")


Студент: Арутюнян Дмитрий (№2)
Ваш персональный SEED: 36390
Запишите SEED в итоговый отчёт — он подтверждает вариант.


In [4]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", 50)
sns.set_theme(style="whitegrid")
plt.rcParams["figure.figsize"] = (8, 4)


In [5]:
#@title Создание персонального датасета (нажмите ▶️) { display-mode: "form" }

def load_it_tickets(seed: int, n: int = 1800) -> "pd.DataFrame":
    """Индивидуальный «грязный» журнал IT-обращений по seed."""
    rng = np.random.default_rng(seed)

    opened_ts = rng.integers(1_735_689_600, 1_751_241_600, size=n)
    department = rng.choice(
        ["Support", "Backend", "DevOps", "Security", "Data"],
        size=n, p=[0.32, 0.22, 0.18, 0.14, 0.14],
    )
    product = rng.choice(
        ["Cloud", "CRM", "Billing", "Auth", "Analytics"], size=n
    )
    priority = rng.choice(
        ["Low", "Medium", "High", "Critical"], size=n, p=[0.40, 0.30, 0.20, 0.10]
    )
    channel = rng.choice(["Email", "Slack", "Portal", "Phone"], size=n)
    assignee = rng.choice(["Junior", "Middle", "Senior"], size=n, p=[0.35, 0.45, 0.20])
    tier = rng.choice(["Free", "Pro", "Enterprise"], size=n, p=[0.45, 0.40, 0.15])
    cpu_pct = np.clip(np.round(rng.normal(48, 18, size=n), 1), 1, 99)
    memory_mb = np.clip(np.round(rng.normal(2200, 850, size=n), 0), 256, None)
    n_reopens = rng.poisson(0.7, size=n)
    comments = rng.poisson(4.0, size=n) + 1
    resolve_hours = np.clip(np.round(rng.lognormal(1.55, 0.75, size=n), 2), 0.2, 240)

    logit = (
        -1.4
        + 1.4 * (priority == "Critical")
        + 0.7 * (priority == "High")
        - 0.4 * (priority == "Low")
        + 0.035 * resolve_hours
        + 0.018 * cpu_pct
        + 0.55 * (department == "Security")
        + 0.35 * (n_reopens >= 2)
        - 0.45 * (assignee == "Senior")
        - 0.30 * (tier == "Enterprise")
    )
    sla = (rng.uniform(size=n) < 1 / (1 + np.exp(-logit))).astype(int)

    df = pd.DataFrame({
        "ticket_id": np.arange(10_000, 10_000 + n),
        "opened_ts": opened_ts,
        "department": department,
        "product": product,
        "priority": priority,
        "channel": channel,
        "assignee_level": assignee,
        "client_tier": tier,
        "cpu_pct": cpu_pct,
        "memory_mb": memory_mb,
        "n_reopens": n_reopens,
        "comments": comments,
        "resolve_hours": resolve_hours,
        "sla_breached": sla,
    })

    df.loc[rng.choice(n, int(0.06 * n), replace=False), "resolve_hours"] = np.nan
    df.loc[rng.choice(n, int(0.04 * n), replace=False), "cpu_pct"] = np.nan
    df.loc[rng.choice(n, int(0.03 * n), replace=False), "comments"] = np.nan
    df.loc[rng.choice(n, 10, replace=False), "resolve_hours"] *= 9
    df.loc[rng.choice(n, 8, replace=False), "memory_mb"] *= 7
    ridx = rng.choice(n, int(0.12 * n), replace=False)
    df.loc[ridx, "department"] = df.loc[ridx, "department"].str.upper()
    pidx = rng.choice(n, int(0.08 * n), replace=False)
    df.loc[pidx, "product"] = "  " + df.loc[pidx, "product"].str.lower() + " "
    df = pd.concat([df, df.sample(int(0.025 * n), random_state=seed)], ignore_index=True)
    return df.sample(frac=1, random_state=seed).reset_index(drop=True)

print("Функция load_it_tickets готова.")


Функция load_it_tickets готова.


In [6]:
df_raw = load_it_tickets(SEED)

print("Ваш датасет создан. Размер:", df_raw.shape)
print("Доля нарушений SLA:", f"{df_raw['sla_breached'].mean():.1%}")


Ваш датасет создан. Размер: (1845, 14)
Доля нарушений SLA: 44.2%


---
# Этап 2. NumPy

**Цель.** Перевести столбец в массив, посчитать агрегаты без цикла `for`, срезать, отфильтровать маской, сделать поэлементную операцию (broadcasting).

Таблица удобна, когда у столбцов есть имена. Массив удобен, когда нужно быстро посчитать «все значения сразу».


### Пример


In [7]:
cpu = df_raw["cpu_pct"].to_numpy(dtype=float)
print("Тип:", type(cpu), "форма:", cpu.shape, "dtype:", cpu.dtype)
print("Первые 8 значений:", cpu[:8])
print("Среднее без учёта NaN:", np.nanmean(cpu))
print("Максимум:", np.nanmax(cpu))

# маска: загрузка выше 80% (NaN даст False)
high_cpu = cpu > 80
print("Доля наблюдений с CPU > 80%:", np.nanmean(high_cpu))

# broadcasting: перевод % в долю 0..1
cpu_share = cpu / 100
print("Доля CPU, первые 5:", cpu_share[:5])


Тип: <class 'numpy.ndarray'> форма: (1845,) dtype: float64
Первые 8 значений: [97.  81.6 47.6 40.6 59.4 60.8 14.6 60.4]
Среднее без учёта NaN: 49.37577639751553
Максимум: 99.0
Доля наблюдений с CPU > 80%: 0.044444444444444446
Доля CPU, первые 5: [0.97  0.816 0.476 0.406 0.594]


### Задание 2

Работайте с массивами, не с циклами `for` по строкам.

1. Соберите массивы `hours` из `resolve_hours` и `mem` из `memory_mb` (тип `float`).
2. Посчитайте **медиану** и **стандартное отклонение** `hours`, игнорируя NaN.
3. Срезом возьмите **последние 15** значений `mem`.
4. Постройте маску «память выше 4000 МБ» и выведите, **сколько** таких тикетов (для NaN считайте «не выше»).
5. Сделайте z-оценку памяти через broadcasting: `(mem - mean) / std`, где mean и std без NaN. Выведите 5 первых z-оценок.


In [8]:
#  ВАШ КОД (Задание 2)

# 1. Массивы hours и mem
hours = df_raw["resolve_hours"].to_numpy(dtype=float)
mem   = df_raw["memory_mb"].to_numpy(dtype=float)

print("hours — тип:", type(hours), "| форма:", hours.shape, "| dtype:", hours.dtype)
print("mem   — тип:", type(mem),   "| форма:", mem.shape,   "| dtype:", mem.dtype)

hours — тип: <class 'numpy.ndarray'> | форма: (1845,) | dtype: float64
mem   — тип: <class 'numpy.ndarray'> | форма: (1845,) | dtype: float64


In [9]:
# 2. Медиана и std для hours

hours_median = np.nanmedian(hours)
hours_std    = np.nanstd(hours)

print(f"Медиана resolve_hours (без NaN): {hours_median:.2f} ч")
print(f"Стандартное отклонение (без NaN): {hours_std:.2f} ч")


Медиана resolve_hours (без NaN): 4.62 ч
Стандартное отклонение (без NaN): 6.73 ч


In [10]:
# 3. Срез последних 15 значений mem
mem_last15 = mem[-15:]

print("Последние 15 значений memory_mb:")
print(mem_last15)

Последние 15 значений memory_mb:
[3312. 2156. 1785. 1954. 2669.  535. 2214. 1861. 3043. 2084. 2342.  459.
 1374. 3760. 2117.]


In [11]:
# 4. Маска памяти > 4000
mask_mem = mem > 4000
n_high_mem = np.sum(mask_mem)
print("Тикетов с памятью > 4000 МБ:", n_high_mem)
print("Доля от всех:", n_high_mem / len(mem))

Тикетов с памятью > 4000 МБ: 42
Доля от всех: 0.022764227642276424


In [22]:
# 5. Z-оценка памяти
mean_mem = np.nanmean(mem)
std_mem  = np.nanstd(mem)
# 5 случайных значений из массива mem
print("5 случайных значений из массива mem:")
print(np.random.choice(mem, size=5, replace=False))

z_mem = (mem - mean_mem) / std_mem
print("Среднее (без NaN):", mean_mem)
print("Std (без NaN):", std_mem)
print("Первые 5 z-оценок:")
print(z_mem[:5])
#  КОНЕЦ ВАШЕГО КОДА


5 случайных значений из массива mem:
[2156. 2143. 2512. 3519. 2044.]
Среднее (без NaN): 2301.3783197831976
Std (без NaN): 1390.9694211544977
Первые 5 z-оценок:
[-0.22960844  1.12843723 -0.52652367 -0.21235429  0.27076201]


**Контрольные вопросы к этапу 2:**

1. Чем массив NumPy принципиально отличается от списка Python?
2. Что такое broadcasting? Приведите пример из этого задания.
3. Почему для среднего по столбцу с пропусками нужен `np.nanmean`, а не `np.mean`?


*Ответы:*

1.Чем массив NumPy принципиально отличается от списка Python?

Тип элементов. Список хранит ссылки на объекты разных типов. Массив NumPy хранит элементы одного типа подряд в памяти.

Скорость. Операции над массивом выполняются векторизованно на уровне C, без интерпретации каждой итерации Python. На миллионе элементов разница — десятки–сотни раз.

Форма. Массив многомерный, фиксированной формы . Список — одномерный; для 2D нужны вложенные списки.

Память. Массив компактнее: хранит значения, а не объекты Python.

Цена. Массив менее гибок: разнотипные данные в один массив не положишь.


2. Что такое broadcasting?

Broadcasting — правило NumPy, которое позволяет делать арифметику над массивами разных, но совместимых форм без явного копирования. Меньший массив «растягивается» по осям большего.

Пример из пункта 5:

z_mem = (mem - mean_mem) / std_mem
mem — массив формы (1845,), mean_mem и std_mem — скаляры. NumPy сам вычитает одно и то же среднее из каждого элемента и делит на одно и то же std — без цикла for.

3.Почему np.nanmean, а не np.mean?

np.mean не игнорирует NaN: по правилам IEEE любое число + NaN = NaN, поэтому результат тоже NaN.

a = np.array([1.0, 2.0, np.nan, 4.0])
np.mean(a)      # nan
np.nanmean(a)   # 2.333...  — NaN пропущен, делим на 3
np.nanmean маскирует пропуски: складывает только валидные значения и делит на их количество.

